### PR TITLE
SONARJAVA-1593 Abstract methods should not be redundant

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  private void checkMethod(MethodTree methodtree) {
+    JavaSymbol.MethodJavaSymbol method = (JavaSymbol.MethodJavaSymbol) methodtree.symbol();
+    JavaSymbol.MethodJavaSymbol overridee = method.overriddenSymbol();
+    if (method.isAbstract() && overridee != null && overridee.owner().isInterface() && !differentContract(method, overridee)) {
+      reportIssue(methodtree, "\"" + method.name() + "\" is defined in the \"" + overridee.owner().name() + "\" interface and can be removed from this class.");
+    }
+  }
+
+  private static boolean differentContract(JavaSymbol.MethodJavaSymbol method, JavaSymbol.MethodJavaSymbol overridee) {
+    return removingParametrizedAspect(method, overridee)
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    return ImmutableList.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+    if (classTree.symbol().isAbstract()) {
+      for (Tree member : classTree.members()) {
+        if (member.is(Tree.Kind.METHOD)) {
+          checkMethod((MethodTree) member);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java]: I would rather check if owner of method is abstract, that would avoid a loop.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+  private static boolean useRawTypeForParametrizedType(Type methodParam, Type overrideeParam) {
+    return overrideeParam instanceof JavaType.ParametrizedTypeJavaType && methodParam.equals(overrideeParam.erasure());
+  }
+
+  private static boolean differentAnnotations(List<JavaSymbol> methodParamSymbols, List<JavaSymbol> overrideeParamSymbols) {
+    for (int i = 0; i < methodParamSymbols.size(); i++) {
+      SymbolMetadataResolve overrideeParamMetadata = overrideeParamSymbols.get(i).metadata();
+      SymbolMetadataResolve methodParamMetadata = methodParamSymbols.get(i).metadata();
+      for (AnnotationInstance annotation : methodParamMetadata.annotations()) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantAbstractMethodCheck.java]: This should probably be factorized.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**